src/core/net/ip6.cpp

```diff
@@ -1199,9 +1235,17 @@ otError Ip6::HandleDatagram(Message &aMessage, Netif *aNetif, const void *aLinkM
         }
 #endif
 
-        IgnoreError(ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse));
+        error = ProcessReceiveCallback(aMessage, messageInfo, nextHeader, aFromNcpHost, Message::kCopyToUse);
+
+        if ((error == OT_ERROR_NONE || error == OT_ERROR_NO_ROUTE) && forward &&
+            !ShouldForwardToThread(messageInfo, aFromNcpHost))
+        {
+            forward = false;
+        }
 
-        SuccessOrExit(error = HandlePayload(aMessage, messageInfo, nextHeader));
+        error =
+            HandlePayload(aMessage, messageInfo, nextHeader, (forward ? Message::kCopyToUse : Message::kTakeCustody));
```

Comment:
SuccessOrExit is not used anymore?

Reply:
To make sure turning off loop even transmit failed.